### PR TITLE
Reword "Pipes and Filters"

### PR DESCRIPTION
**Pull request by rgaiacs**
When preparing for one workshop I thought of reword [Pipes and Filters](https://swcarpentry.github.io/shell-novice/03-pipefilter.html) a little.
## DONE
- [X] Move the files from `data/Users/nelle/data/pdb` to `data/Users/nelle/molecules`.
- [X] Lower the bar of the examples for complete novice learners
- [X] Update the diagram
## TODO
- [ ] Review the exercises (**better leave for another pull request**)

### CODE REVIEW

**Comment by jduckles** [comment]
I'm -1 for writing the files to `/tmp` as that will cause confusion. They're already trying to understand what wildcards **and** redirection mean and you're introducing a new concept `/tmp` at the same time which just distracts from the core concepts of the lesson.  Lets keep the redirects to a file in the current directory, otherwise I like the PR.

**Comment by rgaiacs** [comment]
I addressed @jduckles suggestions.

File: 03-pipefilter.md [hunk 1]
```diff
@@ -16,277 +16,465 @@ Now that we know a few basic commands,
...
+vinyl-chloride.pdb
+vitamin-a.pdb
 ~~~
 
-Let's go into that directory with `cd` and run the command `wc *.pdb`.
-`wc` is the "word count" command:
-it counts the number of lines, words, and characters in files.
-The `*` in `*.pdb` matches zero or more characters,
-so the shell turns `*.pdb` into a complete list of `.pdb` files:
+> ## `.pdb` Extension {.callout}
```
Comment (ChristinaLK) [inline 03-pipefilter.md]: omit this callout?  I'm not sure it's necessary.

File: 03-pipefilter.md [hunk 2]
```diff
@@ -16,277 +16,465 @@ Now that we know a few basic commands,
...
-We can put the sorted list of lines in another temporary file called `sorted-lengths.txt`
-by putting `> sorted-lengths.txt` after the command,
-just as we used `> lengths.txt` to put the output of `wc` into `lengths.txt`.
-Once we've done that,
-we can run another command called `head` to get the first few lines in `sorted-lengths.txt`:
+The `*` in `*nol.pdb` matches zero or more characters,
+so the shell turns `*nol.pdb` into a complete list
+of files that end with `nol.pdf`
+
+> ## Wildcards {.callout}
```
Comment (ChristinaLK) [inline 03-pipefilter.md]: I would move this back to the top, to line 30/77 - we use the wildcard at the very start, so we should explain what it is then too.  One thought: the first paragraph could be outside the callout (as it's necessary information) and the rest of the callout can remain, but be entitled "more about wildcards"

File: 03-pipefilter.md [hunk 3]
```diff
@@ -16,277 +16,465 @@ Now that we know a few basic commands,
...
-~~~
-~~~ {.output}
-  20  156 1158 cubane.pdb
-  12   84  622 ethane.pdb
-   9   57  422 methane.pdb
-  30  246 1828 octane.pdb
-  21  165 1226 pentane.pdb
-  15  111  825 propane.pdb
- 107  819 6081 total
+$ ls *.pdb > content-of-molecules
```
Comment (ChristinaLK) [inline 03-pipefilter.md]: this is detailed, but i'd prefer something shorter than `content-of-molecules`, maybe `molecule-list`?  less to type is better for teaching (and learning)

File: 03-pipefilter.md [hunk 4]
```diff
@@ -16,277 +16,465 @@ Now that we know a few basic commands,
...
+> input. For example, instead of writing `wc ammonia.pdb`, we could write
+> `wc < ammonia.pdb`. In the first case, `wc` gets a command line
+> parameter telling it what file to open. In the second, `wc` doesn't have
+> any command line parameters, so it reads from standard input, but we
+> have told the shell to send the contents of `ammonia.pdb` to `wc`'s
+> standard input.
 
-If we run `wc -l` instead of just `wc`,
-the output shows only the number of lines per file:
+For (2) we will use `wc`
```
Comment (ChristinaLK) [inline 03-pipefilter.md]: Re-explain what (2) is, something like "now we want to be able to count the lines of `content-of-molecules`, as they will tell us how many files were in the directory"

File: 03-pipefilter.md [hunk 5]
```diff
@@ -16,277 +16,465 @@ Now that we know a few basic commands,
...
+And to get only the number of files inside `molecules`
+we can run
 
 ~~~ {.bash}
-$ wc -l *.pdb > lengths.txt
+$ wc -l content-of-molecules
+~~~
+~~~ {.output}
+48
 ~~~
```
Comment (ChristinaLK) [inline 03-pipefilter.md]: Maybe add a concluding sentence like "we've answered our question!  There are 48 pdb files"

File: 03-pipefilter.md [hunk 6]
```diff
@@ -16,277 +16,465 @@ Now that we know a few basic commands,
...
+   54 ethylcyclohexane.pdb
+   54 tuberin.pdb
+   55 vitamin-a.pdb
+   78 cholesterol.pdb
+   79 heme.pdb
+    7 ammonia.pdb
+    9 methane.pdb
 ~~~
 
-The vertical bar between the two commands is called a **pipe**.
```
Comment (ChristinaLK) [inline 03-pipefilter.md]: I like this previous explanation - could we move it up to line 143/163 and then after the "we don't have to know or care," maybe refer people to the "Process" callout, if they DO care how it works.

File: 03-pipefilter.md [hunk 7]
```diff
@@ -16,277 +16,465 @@ Now that we know a few basic commands,
...
+By default, `sort` doesn't now that `1` is the first number,
+`2` is number after `1`, ..., `10` is the number after `9`
+and so on.
+We can request that `sort` interpret `1` as the first number,
+`2` as the number after `1`, ...
+by using the `-n` flag.
 
 ~~~ {.bash}
-$ wc -l *.pdb | sort -n | head -1
+$ wc -l *.pdb | sort
```
Comment (ChristinaLK) [inline 03-pipefilter.md]: need -n.  :)

File: 03-pipefilter.md [hunk 8]
```diff
@@ -16,277 +16,465 @@ Now that we know a few basic commands,
...
+   54 ethylcyclohexane.pdb
+   54 tuberin.pdb
+   55 vitamin-a.pdb
+   78 cholesterol.pdb
+   79 heme.pdb
+  248 lanoxin.pdb
+ 1808 total
 ~~~
 
-This is exactly like a mathematician nesting functions like *log(3x)*
```
Comment (ChristinaLK) [inline 03-pipefilter.md]: I'd like to keep this paragraph as well - right before the Process callout, maybe?

**Comment by ChristinaLK** [comment]
Sorry for leaving this for so long @rgaiacs .  I've made several inline comments; my only other question is: is it possible to truncate the very long output (now that we have lots of files)?  I like that there are more files for data handling purposes, but it's going to make the lesson very tedious to scroll through.  

Also, my comments so far are purely content + organization.  For grammar, do you want me to do inline comments, submit a PR against your branch/repo, or just edit when I merge this in?

**Comment by rgaiacs** [comment]
> I've made several inline comments;

Thanks for all the comments. I will need one-two weeks to work on it.

> my only other question is: is it possible to truncate the very long output (now that we have lots of files)? I like that there are more files for data handling purposes, but it's going to make the lesson very tedious to scroll through.

**Yes.**

> Also, my comments so far are purely content + organization. For grammar, do you want me to do inline comments, submit a PR against your branch/repo, or just edit when I merge this in?

I prefer inline comments so that I will learn when fixing it. =)

**Comment by ChristinaLK** [comment]
That's what I thought, re: grammar.  :) I'll do another round of comments once you've added the changes - no rush!

**Comment by rgaiacs** [comment]
@ChristinaLK I rebased this branch and addressed your comments. Could you review it again?

File: 03-pipefilter.md [hunk 9]
```diff
@@ -16,46 +16,53 @@ Now that we know a few basic commands,
...
+ammonia.pdb
+ascorbic-acid.pdb
+...
+vanillin.pdb
+vinyl-chloride.pdb
+vitamin-a.pdb
 ~~~
 
+The `*` in `*n.pdb` matches zero or more characters,
+so the shell turns `*nol.pdb` into a complete list
```
Comment (lexnederbragt) [inline 03-pipefilter.md]: These three lines mention *n.pdb, *nol.pdb and *.pdb. I'm confused as to whether that is the intention?

File: 03-pipefilter.md [hunk 10]
```diff
@@ -16,46 +16,53 @@ Now that we know a few basic commands,
...
+so lets go into that directory and take a look at that directory:
 
 ~~~ {.bash}
-$ ls molecules
+$ cd molecules
+~~~
+~~~ {.bash}
+$ ls
 ~~~
 ~~~ {.output}
```
Comment (lexnederbragt) [inline 03-pipefilter.md]: When the terminal is wide enough, a simple `ls` will in fact result in two or more columns, as in the original file. Maybe keep that as it is the behavior the learners will experience?

File: 03-pipefilter.md [hunk 11]
```diff
@@ -79,119 +86,86 @@ $ wc *.pdb
...
+"how many files are inside the `molecules` directory?".
+If we could (1) write the output of `ls *.pdb` into a file
+and (2) count the number of lines of that file
+we will have our answer.
+We are very luck and we can do (1) and (2) with the shell.
+For (1) we will use
 
 ~~~ {.bash}
-$ wc -l *.pdb > lengths.txt
+$ ls *.pdb > molecule-list
```
Comment (lexnederbragt) [inline 03-pipefilter.md]: I would suggest using the best-practice of adding extensions also here, so using `molecule-list.txt` as filename

File: 03-pipefilter.md [hunk 12]
```diff
@@ -79,119 +86,86 @@ $ wc *.pdb
 > themselves. It is the shell, not the other programs, that deals with
```
Comment (lexnederbragt) [inline 03-pipefilter.md]: In line 84, `wc`is used, but it has not been used before in your modified lesson. This may confuse learners.

File: 03-pipefilter.md [hunk 13]
```diff
@@ -79,119 +86,86 @@ $ wc *.pdb
...
-$ ls lengths.txt
-~~~
-~~~ {.output}
-lengths.txt
-~~~
+> ## Redirecting Input {.callout}
+>
+> As well as using `>` to redirect a program's output, we can use `<` to
+> redirect its input, i.e., to read from a file instead of from standard
+> input. For example, instead of writing `wc ammonia.pdb`, we could write
```
Comment (lexnederbragt) [inline 03-pipefilter.md]: Again use of `wc` before it is used in the lesson

File: 03-pipefilter.md [hunk 14]
```diff
@@ -203,90 +177,285 @@ or copy data from one program to the other in memory,
...
+   54 ethylcyclohexane.pdb
+   54 tuberin.pdb
+   55 vitamin-a.pdb
+   78 cholesterol.pdb
+   79 heme.pdb
+    7 ammonia.pdb
+    9 methane.pdb
+~~~
+
+Ops. Something went catastrophic wrong.
```
Comment (lexnederbragt) [inline 03-pipefilter.md]: On my filesystem the sort operation works fine, due to the extra spaces the `wc` command puts in from of the numbers. The result is

```
$ wc -l *.pdb | sort
       7 ammonia.pdb
       9 methane.pdb
      10 methanol.pdb
      10 vinyl-chloride.pdb
      12 ethane.pdb
....
      78 cholesterol.pdb
      79 heme.pdb
     248 lanoxin.pdb
    1808 total
```

 This has been a thing I didn't like about the lesson before, and it hops over the need for `sort -n` when sorting numbers...

File: 03-pipefilter.md [hunk 15]
```diff
@@ -79,119 +86,86 @@ $ wc *.pdb
...
-If you think this is confusing,
-you're in good company:
-even once you understand what `wc`, `sort`, and `head` do,
-all those intermediate files make it hard to follow what's going on.
-We can make it easier to understand by running `sort` and `head` together:
+each time that you want to know the number of files inside a directory
+will be tedious.
+If we could drop the file our life will be more easy.
+Again,
+we are luck and we can acomplish it with
```
Comment (lexnederbragt) [inline 03-pipefilter.md]: acomplish -> accomplish

**Comment by lexnederbragt** [comment]
in `fig/redirects-and-pipes.png`, the bottom command seems mangled, maybe two commands in the same place?

**Comment by wking** [comment]
Adding to [Lex's comment][1], the redirects-and-pipes image also still has `wc` as the first step in the middle (redirect to file) entry, but the command-line version there is now talking about `ls`.

[1]: https://github.com/swcarpentry/shell-novice/pull/225#issuecomment-151509082

**Comment by gdevenyi** [comment]
Hi, this needs a rebase to be reviewed. Thanks

File: 03-pipefilter.md [hunk 16]
```diff
@@ -16,46 +16,53 @@ Now that we know a few basic commands,
 we can finally look at the shell's most powerful feature:
 the ease with which it lets us combine existing programs in new ways.
 We'll start with a directory called `molecules`
-that contains six files describing some simple organic molecules.
-The `.pdb` extension indicates that these files are in Protein Data Bank format,
-a simple text format that specifies the type and position of each atom in the molecule.
+so lets go into that directory and take a look at that directory:
```
Comment (shwina) [inline 03-pipefilter.md]: "Let's start by going into the `molecules` directory and having a look at what's in there"

**Comment by gvwilson** [comment]
@rgaiacs please rebase

**Comment by rgaiacs** [comment]
Thanks for all the review. I'm dropping this pull request since will take a long time for me to rebase this.
